### DISCUSSION TITLE
Custom MDAEngine: how to integrate devices with no device adapters

### DISCUSSION THREAD

**Discussion by jacopoabramo** [pymmcore-plus/pymmcore-plus#395] [Q&A] [answered]
This [section](https://pymmcore-plus.github.io/pymmcore-plus/guides/custom_engine/) of the `pymmcore-plus` guide states that writing a custom acquisition engine allows to control devices that do not have a device adapter for micro-manager; could someone maybe provide an example of a device that can be integrated in a custom MDAEngine? I'm interested in understanding if there's a specific set of methods to declare or if it's completely up to the user and maybe a suggested standard approach.

**Comment by tlambert03** [comment]
Thanks for the question @jacopoabramo.

it's completely up to the user at this point.  One thing I'm rather hesitant to do is to introduce yet another device interface.

I see a couple standard use cases:

1. Someone is *mostly* content using micro-manager devices, but wants to add one-off control of something else (say, a micro-fluidics device). For that, many people [subclass the MDAEngine](https://pymmcore-plus.github.io/pymmcore-plus/guides/custom_engine/#customizing-the-mdaengine) and customize it slightly, either to do additional setup before executing a "regular" event (as micromanager would have done), or to fully defer to some unusual event (for example, based on [custom metadata](https://pymmcore-plus.github.io/pymmcore-plus/guides/custom_engine/#utilizing-mdaevent-metadata) that someone has added to the event definition).  This is targeted at tinkerers/scientists who want full control over a very small additional part of the setup/exec flow.
2. Someone wants to combine micro-manager with some other platform that has declared a new device interface (say, python-microscope, imswitch, navigate, etc...).  I don't have code to give you there, since I'm not that familiar with the exact invocation of all the other device interfaces  but essentially it would still just boil down to subclassing `MDAEngine` and overriding `setup_event` and/or `exec_event` to direct the newly controlled interface according to the intentions of the MDAEvent.
3. Someone doesn't want micromanager at all, but does want to take advantage of the Runner, which dispatches a sequence of MDAEvent objects to some new concrete implementation of the [engine protocol](https://pymmcore-plus.github.io/pymmcore-plus/guides/custom_engine/#the-mdaengine-protocol).  This is the part where it might not be clear that micro-manager needn't be involved at all (see https://github.com/pymmcore-plus/useq-schema/pull/197/files for a PR in useq-schema that would extract just the runner logic from pymmcore-plus into a micromanager-agnostic availability.  There, any library that implements device control can just worry about translation the intention (the `MDAEvent`) into execution, by dispatching devices accordingly.

i know that's still rather high level. So let me know if you're interested in a slightly more specific use case, and I can try to give more details

**Reply by jacopoabramo** [reply]
@tlambert03 thanks for the extensive reply!

> One thing I'm rather hesitant to do is to introduce yet another device interface.

That's exactly what I'm afraid as well, that's why the question in the first place. My goal would be achieving point 2; I'm currently not targeting a specific backend, so for example let's assume that I have a class representing a motor via an imaginary protocol:

```py
class CustomMotor:
    
    def __init__(self, axes: list[str]) -> None:
         self._position: Tuple[int | float, ...] = (0, ...)
         ... # initializing the motor according to it's DLL/python package/whateves
    
    def move(self, axis: str, pos: float | int) -> Optional[Status]:
         # do my move operation here;
         # an (again, imaginary) status object returns the condition of the operation;
         # or alternatively it doesn't return anything at all
         ...
    
    def go_home(self) -> Optional[Status]:
         # same as above, all assumptions
         ...
     
    def shutdown(self) -> None:
         # cleanup operation
         ...
     
    @property
    def position(self) -> Tuple[int | float, ...]
         # assuming the possibility of having a property that returns the current position
         return self._position
```

What I'm interested in is how would I treat the implementation of a custom `MDAEngine` protocol with a `CustomMotor` protocol like this. I understand that it can be a lot to ask to make such speculation on some lose information but I'm trying to understand the logic behind how customization works.

EDIT: rereading your reply, I should clarify that with "`MDAEngine` protocol" I mean subclassing it since like you said it keeps the compatibility with the existing handling of micro-manager devices. Whoopsies

**Comment by tlambert03** [comment, marked as answer]
yeah that's a good example. 

the main thing you would need to do is to translate the intention of a useq [`MDAEvent`](https://pymmcore-plus.github.io/useq-schema/schema/event/) into some action taken by your device.  So, let's say you want to use micromanager for everything but some custom stage.  The simplest integration would be:

```python
from pymmcore_plus import CMMCorePlus
from pymmcore_plus.mda import MDAEngine
import useq


class MyEngine(MDAEngine):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._xy_motor = CustomMotor()

    def setup_event(self, event: useq.MDAEvent) -> None:
        """Prepare state of system (hardware, etc.) for `event`."""
        super().setup_event(event)
        # move to the requested position using our motor
        self._xy_motor.move(event.x_pos, event.y)


# load the mm-part
core = CMMCorePlus.instance()
core.loadSystemConfiguration()

# Register the custom engine with the runner
custom_engine = MyEngine(core)
core.mda.set_engine(custom_engine)

# Run an MDA
seq = useq.MDASequence(
    time_plan=useq.TIntervalLoops(interval=1, loops=10),
    stage_positions=[(0, 0), (1, 1)],
)
core.mda.run(seq)
```

naturally, if you start to have multiple devices, the engine itself starts to get heavy. and that may be fine in some cases, but if you would like to give me an example of how to programmatically use imswitch to create something like a core object that controls a handful of devices, I could create a more elaborate example that splits control of devices across the two frameworks

**Reply by jacopoabramo** [reply]
Perfect, that's what I generally needed. Thanks!

EDIT: I actually have a couple of more questions related to the engine but for archival purposes I'll create a separate discussion for each one.

**Reply by tlambert03** [reply]
sounds good!

another side note, geared towards you as someone who thinks about this stuff a lot.  that `useq.MDAEvent` obviously takes on a lot of importance in this pattern.  It's the declarative object that somehow needs to capture all the stuff that anyone might want to execute with a concrete engine.  It's the "what" and the engine is the "how".  So, by all means, if you continue to look at this pattern and hit what feels to be a limitation in the `MDAEvent` design, please open an issue at useq-schema to discuss how that object could be modified to meet more use cases.  It's obviously somewhat opinionated (in that it acknowledges a 3D space, the concept of time, the concept of a detector that may have exposure)... and that's attempting to be a balance between making "common things easy" and total abstraction.  There are some places to put arbitrary information, such as `metadata` and `properties`... but that's specifically an object that I would welcome feedback on

**Comment by tlambert03** [comment]
sorry... my example used your `CustomMotor.move` API incorrectly :joy: ... but hopefully you get the point  I guess I should have created two motors and called `x_motor.move('x', event.x_pos)`, etc...

**Comment by tlambert03** [comment]
and just for the sake of completeness... these two lines in my example that access the `CMMCorePlus.mda` attribute could themselves be done without the micromanager core object.

```python
core.mda.set_engine(custom_engine)
core.mda.run(seq)
```

those are just there as a convenience to instantiate `MDARunner`.  It could be replaced with:

```python
runner = MDARunner()
runner.set_engine(custom_engine)
runner.run(seq)
```

and that's specifically the object that could be fully extracted from this library, as is done in https://github.com/pymmcore-plus/useq-schema/pull/197

**Reply by jacopoabramo** [reply]
Hm, could you maybe clarify the implication of setting a different custom engine via the runner? Would this mean that I lose access to the core?

**Reply by tlambert03** [reply]
if you subclass and extend `pymmcore_plus.mda.MDAEngine`, then no, you don't lose access to the core.  Note that [`pymmcore_plus.mda.MDAEngine` expects an instance of `CMMCorePlus` in its constructor].  It is specifically the object that implements the abstract engine protocol expected by the Runner to translate `MDAEvents` into specific actions taken by a CMMCore instance. 

the main point was just that `CMMCorePlus.mda` is just a plain old instance of `MDARunner` (provided as a convenience for `CMMCorePlus` users).  But the runner itself knows nothing about the core (only the "engine" that the runner drives does).
